core/kernel/msg_param.c

```diff
@@ -31,6 +31,121 @@
 #include <kernel/msg_param.h>
 #include <mm/mobj.h>
 
+/**
+ * msg_param_extract_pages() - extract list of pages from
+ * OPTEE_MSG_ATTR_NONCONTIG buffer.
+ *
+ * @buffer:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_pages:  number of pages in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ *
+ * You can think that @buffer points to PA of the following structure:
+ *
+ * struct page_data {
+ *   uint64_t pages_array[OPTEE_MSG_NONCONTIG_PAGE_SIZE/sizeof(uint64_t) - 1];
+ *   uint64_t next_page_data;
+ * };
+ *
+ * So, it is like linked list of arrays, where every element of linked list fits
+ * exactly into one 4K page.
+ *
+ * This function extracts data from arrays into one array pointed by @pages
+ *
+ * @buffer points to data shared with NW, so some precautions should be taken.
+ */
+static ssize_t msg_param_extract_pages(paddr_t buffer, paddr_t *pages,
+				       size_t num_pages)
+{
+	size_t cnt;
+	struct mobj *mobj;
+	paddr_t page;
+	uint64_t *va;
+
+	if (buffer & SMALL_PAGE_MASK)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/*
+	 * There we map first page of array.
+	 * mobj_mapped_shm_alloc() will check if page resides in nonsec ddr
+	 */
+	mobj = mobj_mapped_shm_alloc(&buffer, 1, 0, 0);
+	if (!mobj)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	va = mobj_get_va(mobj, 0);
+	assert(va);
+
+	for (cnt = 0; cnt < num_pages; cnt++, va++) {
+		/*
+		 * If we about to roll over page boundary, then last entry holds
+		 * address of next page of array. Unmap current page and map
+		 * next one
+		 */
+		if (!((vaddr_t)(va + 1) & SMALL_PAGE_MASK)) {
+			page = *va;
+
+			mobj_free(mobj);
+
+			if (page & SMALL_PAGE_MASK)
+				return TEE_ERROR_BAD_PARAMETERS;
+
+			mobj = mobj_mapped_shm_alloc(&page, 1, 0, 0);
+			if (!mobj)
+				return TEE_ERROR_BAD_PARAMETERS;
+
+			va = mobj_get_va(mobj, 0);
+			assert(va);
+		}
+		pages[cnt] = *va;
+	}
+
+	mobj_free(mobj);
+	return cnt;
+}
+
+struct mobj *msg_param_mobj_from_noncontig(const struct optee_msg_param *param,
+					   bool map_buffer)
+{
+	struct mobj *mobj = NULL;
+	paddr_t *pages;
+	paddr_t page_offset;
+	ssize_t num_pages, pages_cnt;
+	uint64_t buf_ptr;
+
+	assert(param->attr & OPTEE_MSG_ATTR_NONCONTIG);
+
+	/* Make sure that NW will not change value in SHM */
+	buf_ptr = param->u.tmem.buf_ptr;
+
+	page_offset = buf_ptr & SMALL_PAGE_MASK;
+	num_pages = (param->u.tmem.size + page_offset - 1)
+		/ SMALL_PAGE_SIZE + 1;
```

Comment:
Please put the `/` at the end of the line above and align this row with the `(` on the row above.